pyswarms/base/base_discrete.py

```diff
@@ -81,13 +81,16 @@ def __init__(
             a tuple of size 2 where the first entry is the minimum velocity
             and the second entry is the maximum velocity. It
             sets the limits for velocity clamping.
-        ftol : float, optional
-            relative error in objective_func(best_pos) acceptable for
-            convergence. Default is :code:`-np.inf`.
+        ftol : float
+            relative error in objective_func(best_pos) acceptable for convergence.
+            To deactivate it in order to process all iterations, use any negative value.
+            Deactivating ftol also disables ftol_iter.
+            Default is :code: `-np.inf` to disable ftol
         ftol_iter : int
-            number of iterations over which the relative error in
-            objective_func(best_pos) is acceptable for convergence.
-            Default is :code:`1`
+            number of consecutive iterations over which the relative change in
+            objective_func(best_pos) is stalled or less than ftol. It works
+            when ftol is greater than zero (e.g. 1e-6)
+            Default is :code:`50`
```

Comment:
I'm not exactly convinced how `ftol_iter` can work for both the **acceptable tolerance** and **stall** cases. Maybe I'm missing something here.

The stall condition is when there is no improvement in the optimal value over `n_iter_no_change` (`stall_iter`) iterations. The `ftol_iter` should be checking that the `ftol` threshold must be achieved for the `ftol_iter` consecutive iterations.

Ill try to draw up an example but just for a sanity check (minimization problem)
```py
.
19: 50
20: 50    # <- stalled for 20 iterations (stop here)
21: 50    # <- if no stall iter, keep going as ftol is not achieved (will run all the iterations)
.
```
@ljvmiranda921 Are we on the same page here?

Reply:
@nishnash54 , you are missing what I was missing. With the current implementation, you never reach to iteration #21 with ftol_iter=1. You are stopped after 1st or 2nd iteration.

Please check the next post.